client/desktop/lib/notifications/api/notes.js

```diff
@@ -0,0 +1,71 @@
+/*
+ * Internal dependencies
+ */
+const state = require( 'desktop/lib/state' );
+const handler = require( 'wpcom-xhr-request' );
+
+const promiseTimeout = function ( ms, promise ) {
+	const timeout = new Promise( ( _, reject ) => {
+		const id = setTimeout( () => {
+			clearTimeout( id );
+			reject( `Request timed out in ${ ms } ms` );
+		}, ms );
+	} );
+
+	return Promise.race( [ promise, timeout ] );
+};
+
+async function fetchNote( noteId ) {
+	return new Promise( ( resolve, reject ) => {
+		handler(
+			{
+				path: `/notifications/${ noteId }`,
+				authToken: state.getUser().token,
+				apiVersion: '1.1',
+				query: {
+					fields: 'id,type,unread,body,subject,timestamp,meta,note_hash',
+				},
+			},
+			( error, body ) => {
+				if ( error ) {
+					reject( error );
+				}
+
+				if ( body.notes && Array.isArray( body.notes ) && body.notes.length > 0 ) {
+					return resolve( body.notes[ 0 ] );
+				}
+
+				return resolve( null );
+			}
+		);
+	} );
+}
+
+async function markReadStatus( noteId, isRead ) {
+	return new Promise( ( resolve, reject ) => {
+		handler(
+			{
+				method: 'POST',
+				path: '/notifications/read',
+				apiVersion: '1.1',
+				authToken: state.getUser().token,
+				body: {
+					counts: {
+						[ noteId ]: isRead ? 9999 : -1, // magic values required by the API \_(ツ)_/¯
+					},
+				},
+			},
+			( error ) => {
+				if ( error ) {
+					reject( error );
+				}
+				resolve( null );
+			}
+		);
+	} );
+}
+
+module.exports = {
+	fetchNote: ( noteId ) => promiseTimeout( 300, fetchNote( noteId ) ),
+	markReadStatus: ( noteId, isRead ) => promiseTimeout( 1000, markReadStatus( noteId, isRead ) ), // response time of this endpoint is very slow!
```

Comment:
Why is there only a 0.3s timeout for `fetchNote`? That's very unfriendly for users on slow connections. And why do we need hard timeouts at all?

Reply:
If the `fetchNote` or `markReadStatus` request takes a lot of time, it doesn't block the UI. It's only that the incoming WebSocket notifications take long to appear in the UI, or that marking a notification as read takes a long time or is never really performed on server. The user can continue to use the UI normally.

Nowhere else in Calypso do we put extra timeouts on REST APIs. And even the 2 seconds still are way too short for, e.g., users in Australia or South Africa who are connecting to US servers.

I'd remove them everywhere except the initial boot sequence, which has a potential to brick the UI completely.